orttraining/orttraining/core/optimizer/memory_alleviation.h

```diff
@@ -0,0 +1,285 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include <charconv>
+#include "core/common/inlined_containers.h"
+#include "core/common/string_utils.h"
+#include "core/optimizer/graph_transformer.h"
+
+namespace onnxruntime {
+
+/**
+@Class MemoryAlleviation
+
+Find recomputable subgraphs and enable according to user configs.
+*/
+
+class MemoryAlleviation : public GraphTransformer {
+ private:
+  using NodeOutputPort = std::pair<const Node*, int>;
+  using ActivationUsedMap = InlinedHashMap<std::string, std::pair<bool, bool>>;
+
+  /**
+   * @brief Level to control allowed operations during subgraph detecting.
+   * Level 0: only allow cheap-to-compute operations.
+   * Level 1: allow more expensive operations.
+   */
+  enum class ProbeLevel {
+    Basic = 0,
+    Advanced = 1,
+  };
+
+  /**
+   * @brief Type of memory reduction techniques.
+   */
+  enum class AlleviationType {
+    None = 0,  // Disabled.
+    Recompute = 1,
+  };
+
+  /**
+   * @brief Type of user config.
+   * type: type of memory reduction techniques.
+   * requested_count: the number of occurrences of a subgraph pattern for alleviation. -1 means apply all.
+   *   One example: if a subgraph pattern is found 3 times, and requested_count is set 2, then the 1st and 2nd subgraph in
+   *   typological order will be applied for alleviation. This is useful to avoid alleviating more memory than needed.
+   */
+  struct UserAlleviationConfig {
+    AlleviationType type;
+    int requested_count;
+  };
+
+  struct EntryOperatorConfig {
+    InlinedVector<int> input_arg_indices;  // input index to iterate further (bottom up)
+  };
+
+  /**
+   * @brief Struct to store properties of a specific subgraph.
+   */
+  struct AlleviationSubGraphDesc {
+    AlleviationSubGraphDesc() = default;
+
+    // A string to represent the subgraph, used as a unique "ID" for a unique subgraph.
+    std::string subgraph_representative_str;
+
+    InlinedHashMap<std::string, int> shape_str_frequency;  // shape string to frequency
+    UserAlleviationConfig user_alleviation_config;
+    int total_frequency{0};  // The occurrence of this subgraph pattern in the graph.
+
+    int applied_count{0};  // The number of times this subgraph pattern has been really applied in this transformer.
+    int skip_count{0};     // The number of times this subgraph instances will skipped in reversed topological order.
```

Comment:
If we have total_frequency and applied_count then do we need skip_count?

Reply:
Good question. Yes, we need. We iterated the graph in revered topo order. So we need know how many sub graph already skipped during the graph modification. 

For the below code:
`subgraph_desc.total_frequency - user_config.requested_count` is used to calculate the intended skip count, we still need a skip_count to store how many we already skips.
```
  int skip_count = (user_config.requested_count == -1)
                       ? 0
                       : std::max(0, subgraph_desc.total_frequency - user_config.requested_count);

  subgraph_desc.skip_count += 1;

  if (user_config.type != OptimizationType::None && subgraph_desc.skip_count > skip_count) 
```